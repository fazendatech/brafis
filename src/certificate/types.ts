/**
 * @description Opções para configurar um certificado.
 *
 * @property {Buffer} pfx - Dados PFX (Personal Information Exchange) para o certificado.
 * @property {string} password - Senha para descriptografar os dados PFX.
 */
export interface CertificateP12Options {
  pfx: Buffer;
  password: string;
}

/**
 * @description Representa o formato PEM (Privacy-Enhanced Mail).
 *
 * @property {string} cert - Dados PEM (Privacy-Enhanced Mail) para o certificado.
 * @property {string} key - Chave privada armazenada com os dados PFX.
 */
export interface PemPayload {
  cert: string;
  key: string;
}

/**
 * @description Representa o formato PFX (PKCS#12).
 *
 * @property {Uint8Array} raw - Os dados PFX (Personal Information Exchange).
 * @property {string} password - A senha para descriptografar os dados PFX.
 */
export interface P12Payload {
  raw: Uint8Array;
  password: string;
}

/**
 * @description Representa os campos de um certificado.

 * @property {Record<string, string>} subject - Dados do requerente do certificado
 * @property {Record<string, string>} issuer - Dados do emissor do certificado
 * @property {Date} validFrom - A data a partir da qual o certificado é válido.
 * @property {Date} validTo - A data até a qual o certificado é válido.
 * @property {string} serialNumber - O número de série do certificado.
 * @property {string} signatureAlgorithm - O algoritmo usado para assinar o certificado. E.g. `sha256WithRSAEncryption`, `rsaEncryption`, ...
 */
export interface CertificateFields {
  subject: Record<string, string>;
  issuer: Record<string, string>;
  validFrom: Date;
  validTo: Date;
  serialNumber: string;
  signatureAlgorithm: string;
}

export interface CertificateP12AsPemOptions {
  allowExpired?: boolean;
}
