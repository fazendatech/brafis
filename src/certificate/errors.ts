/**
 * @description Não foi possível extrair conteúdo do arquivo PFX.
 */
export class InvalidPfxError extends Error {
  constructor() {
    super("PFX must have valid certificate content.");
    this.name = "InvalidPfxError";
  }
}

/**
 * @description Senha inválida para o arquivo PFX
 */
export class InvalidPasswordError extends Error {
  constructor() {
    super("Invalid password.");
    this.name = "InvalidPasswordError";
  }
}

/**
 * @description Não foi possível encontrar uma chave privada no conteúdo do arquivo.
 */
export class NoPrivateKeyFoundError extends Error {
  constructor() {
    super("No private key found in the PFX file.");
    this.name = "NoPrivateKeyFoundError";
  }
}

/**
 * @description Não foi possível encontrar um certificado válido no conteúdo do arquivo.
 */
export class NoCertificatesFoundError extends Error {
  constructor() {
    super("No valid certificates found in the PFX file.");
    this.name = "NoCertificatesFoundError";
  }
}

/**
 * @description Certificado expirado.
 */
export class CertificateExpiredError extends Error {
  constructor() {
    super("Certificate is expired.");
    this.name = "CertificateExpiredError";
  }
}
