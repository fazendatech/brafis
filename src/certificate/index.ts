import forge from "node-forge";

/**
 * @description Interface que representa os campos extraídos de um certificado PEM.
 */
interface CertificateFields {
  subject: { name: string; value: string }[];
  issuer: { name: string; value: string }[];
  validFrom: Date;
  validTo: Date;
  serialNumber: string;
  publicKey: string;
  signatureAlgorithm: string;
}

/**
 * @description Interface que representa o formato PEM (Privacy-Enhanced Mail).
 * É usada para armazenar tanto o certificado quanto a chave privada no formato PEM.
 * PEM é um formato codificado em base64 comumente usado para certificados e chaves.
 */
interface PEM {
  cert: string;
  key: string;
}

/**
 * @description Interface que representa o formato PFX (PKCS#12).
 * Ela contém os dados do PFX em uma string codificada em base64 (`bufferString`)
 * e a senha (`pass`) usada para descriptografar a chave privada.
 */
interface P12 {
  bufferString: string;
  pass: string;
}

/**
 * @description Interface que representa um "bag" de certificado usado para armazenar certificados.
 * Este "bag" contém um objeto `forge.pki.Certificate`, que possui informações
 * sobre o certificado, como o sujeito, emissor, validade e chave pública.
 */
interface CertBag {
  cert: forge.pki.Certificate;
}

/**
 * @description Classe que representa um Certificado que pode converter arquivos PFX (PKCS#12) para o formato PEM.
 * Ela fornece métodos para extrair certificados e chaves privadas de arquivos PFX e
 * convertê-los para o formato PEM. O formato PEM é frequentemente usado para importar/exportar certificados
 * e chaves em ambientes como servidores web ou serviços em nuvem.
 */
export class Certificate {
  private readonly pfxData: P12;
  private pemData?: PEM;
  private certFields?: CertificateFields;

  /**
   * @description Getter para obter os campos do Certificado.
   * Se os campos do Certificado ainda não estiverem populados, ele aciona a extração do certificado PEM.
   *
   * @returns {CertificateFields} Um objeto contendo os campos do certificado extraídos do certificado PEM.
   */
  public get pemFields(): CertificateFields {
    if (!this.certFields) {
      this.extractPemFields();
      if (!this.certFields) {
        throw new Error("Certificate fields are not available.");
      }
    }
    return this.certFields;
  }

  /**
   * @description Getter para obter os dados PEM (certificado e chave privada).
   * Se os dados PEM ainda não estiverem populados, ele aciona a conversão de PFX para PEM.
   *
   * @returns {PEM} Um objeto contendo o certificado e a chave privada no formato PEM.
   */
  get pem(): PEM {
    if (!this.pemData) {
      this.p12ToPem();
      if (!this.pemData) {
        throw new Error("PEM data is not available.");
      }
    }
    return this.pemData;
  }

  /**
   * @description Construtor da classe para inicializar a instância do Certificado com um arquivo PFX e uma senha.
   * O arquivo PFX é fornecido como um `ArrayBuffer`, e a senha é usada para desbloquear
   * a chave privada contida no arquivo PFX.
   *
   * @param {ArrayBuffer} pfx - O buffer contendo os dados do arquivo PFX, normalmente de um arquivo.
   * @param {string} passphrase - A senha usada para descriptografar o arquivo PFX e extrair a chave privada.
   */
  constructor(pfx: ArrayBuffer, passphrase: string) {
    const pfxUint8Array = new Uint8Array(pfx);
    this.pfxData = {
      bufferString: forge.util.binary.base64.encode(pfxUint8Array),
      pass: passphrase,
    };
  }

  /**
   * @description Converte um arquivo PFX (PKCS#12) para o formato PEM.
   * Este método decodifica os dados PFX de base64 e extrai a chave privada e os certificados.
   * Em seguida, classifica os certificados por data de expiração, retornando o certificado com a maior
   * validade e sua respectiva chave privada no formato PEM. Se nenhum certificado ou chave privada
   * for encontrado no arquivo PFX, um erro é lançado.
   *
   * @returns {PEM} Um objeto contendo o certificado e a chave privada no formato PEM.
   *
   * @throws {Error} Se nenhum certificado ou chave privada for encontrado no arquivo PFX.
   *
   * @private
   */
  private p12ToPem(): PEM {
    // Decode the base64-encoded PFX data
    const p12Der = forge.util.decode64(this.pfxData.bufferString);
    // Convert the DER-encoded PFX data into an ASN.1 structure
    const p12Asn1 = forge.asn1.fromDer(p12Der);
    // Parse the ASN.1 structure into a PKCS#12 object
    const p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, true, this.pfxData.pass);

    const oids = forge.pki.oids;
    // Extract the encrypted and unencrypted private key data
    const keyBags = p12.getBags({
      bagType: oids.pkcs8ShroudedKeyBag,
    })[oids.pkcs8ShroudedKeyBag];
    const unencryptedKeyBags = p12.getBags({ bagType: oids.keyBag })[
      oids.keyBag
    ];
    const keyData = (keyBags || []).concat(unencryptedKeyBags || []);

    // Extract the certificates from the PFX and sort them by expiration date
    const certBags: CertBag[] = p12.getBags({
      bagType: oids.certBag,
    })[oids.certBag] as CertBag[];

    if (!certBags || certBags.length === 0) {
      throw new Error("No certificates found in the PFX file.");
    }

    // Sort certificates by expiration date (oldest first)
    certBags.sort((a: CertBag, b: CertBag) => {
      return (
        new Date(a.cert.validity.notAfter).getTime() -
        new Date(b.cert.validity.notAfter).getTime()
      );
    });

    // Throw an error if no private key is found
    if (!keyData[0]?.key) {
      throw new Error("No private key found in the PFX file.");
    }

    const pki = forge.pki;
    // Wrap the first private key into an RSA Private Key info structure
    const rsaPrivateKey = pki.privateKeyToAsn1(keyData[0].key);
    const privateKeyInfo = pki.wrapRsaPrivateKey(rsaPrivateKey);

    // Convert the first certificate and private key to PEM format
    const cert = pki.certificateToPem(certBags[0].cert);
    const key = pki.privateKeyInfoToPem(privateKeyInfo);

    // Store the PEM data (certificate and private key)
    this.pemData = { cert, key };

    // Return the certificate and private key in PEM format
    return this.pemData;
  }

  /**
   * @description Extrai campos do certificado PEM.
   * Este método pega um certificado PEM e extrai campos importantes como
   * sujeito, emissor, período de validade, número de série, chave pública e algoritmo de assinatura.
   *
   * @returns {CertificateFields} Um objeto contendo os campos extraídos do certificado.
   */
  private extractPemFields(): CertificateFields {
    const pki = forge.pki;
    const cert = pki.certificateFromPem(this.pem.cert);

    this.certFields = {
      subject: cert.subject.attributes.map((attr) => ({
        name: attr.name || "",
        value: typeof attr.value === "string" ? attr.value : "",
      })),
      issuer: cert.issuer.attributes.map((attr) => ({
        name: attr.name || "",
        value: typeof attr.value === "string" ? attr.value : "",
      })),
      validFrom: cert.validity.notBefore,
      validTo: cert.validity.notAfter,
      serialNumber: cert.serialNumber,
      publicKey: forge.util.decodeUtf8(pki.publicKeyToPem(cert.publicKey)),
      signatureAlgorithm: cert.siginfo.algorithmOid,
    };

    return this.certFields;
  }
}
