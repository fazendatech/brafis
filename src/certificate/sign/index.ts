import { SignedXml } from "xml-crypto";

import type { CertificateP12 } from "@/certificate";
import { makeBuilder, makeParser } from "@/utils/xml";

/**
 * @description Opções para assinar objetos XML.
 *
 * @property {CertificateP12} certificate - O certificado digital no formato PKCS#12 usado para assinar
 * @property {XmlObject} xml - O conteúdo do documento XML a ser assinado
 * @property {string} signId - O ID do elemento a ser assinado
 */
export interface signXmlOptions<XmlObject> {
  certificate: CertificateP12;
  xmlObject: XmlObject;
  signId: string;
}

/**
 * @description Retorna o objeto XML com assinatura.
 *
 * @param {signXmlOptions} - Opções para assinar objetos XML.
 *
 * @returns Objeto XML assinado.
 */
export function signXml<XmlObject>({
  certificate,
  xmlObject,
  signId,
}: signXmlOptions<XmlObject>) {
  const xml = makeBuilder().build(xmlObject);

  const { key, cert } = certificate.asPem();
  const sig = new SignedXml({
    privateKey: key,
    publicCert: cert,
    signatureAlgorithm: "http://www.w3.org/2000/09/xmldsig#rsa-sha1",
    canonicalizationAlgorithm:
      "http://www.w3.org/TR/2001/REC-xml-c14n-20010315",
  });

  sig.addReference({
    xpath: `//*[@Id='${signId}']`,
    transforms: [
      "http://www.w3.org/2000/09/xmldsig#enveloped-signature",
      "http://www.w3.org/TR/2001/REC-xml-c14n-20010315",
    ],
    digestAlgorithm: "http://www.w3.org/2000/09/xmldsig#sha1",
  });

  sig.computeSignature(xml);
  const signedXml = sig.getSignedXml();
  return makeParser().parse(signedXml);
}
