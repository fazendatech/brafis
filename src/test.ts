import { CertificateP12 } from "@/certificate";
import { NfeWebServices } from "@/dfe/nfe/webServices";

const certificate = await CertificateP12.fromFilepath({
  filepath: process.env.TEST_CERTIFICATE_PATH ?? "",
  password: process.env.TEST_CERTIFICATE_PASSWORD ?? "",
});

const service = new NfeWebServices({
  uf: "DF",
  env: "homologacao",
  certificate: certificate,
});

const statusServ = await service.statusServico();

// biome-ignore lint/suspicious/noConsoleLog: This is a test file
// biome-ignore lint/suspicious/noConsole: This is a test file
console.log(JSON.stringify(statusServ, null, 2));

const consCad = await service.consultaCadastro({
  CPF: process.env.TEST_CPF ?? "",
});
// biome-ignore lint/suspicious/noConsoleLog: This is a test file
// biome-ignore lint/suspicious/noConsole: This is a test file
console.log(JSON.stringify(consCad, null, 2));

const consProt = await service.consultaProtocolo({
  chNFe: process.env.TEST_CHNFE ?? "",
});
// biome-ignore lint/suspicious/noConsoleLog: This is a test file
// biome-ignore lint/suspicious/noConsole: This is a test file
console.log(JSON.stringify(consProt, null, 2));

const dfe = await service.distribuicaoDfe({
  CPF: process.env.TEST_CERTIFICATE_CPF ?? "",
  consChNFe: { chNFe: process.env.TEST_CHNFE ?? "" },
});
// biome-ignore lint/suspicious/noConsoleLog: This is a test file
// biome-ignore lint/suspicious/noConsole: This is a test file
console.log(JSON.stringify(dfe, null, 2));
