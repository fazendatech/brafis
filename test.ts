import { CertificateP12 } from "@/certificate";
import { NfeWebServices } from "@/dfe/nfe/webServices";

const certificate = await CertificateP12.fromFilepath({
  filepath: process.env.TEST_CERTIFICATE_PATH ?? "",
  password: process.env.TEST_CERTIFICATE_PASSWORD ?? "",
});

const service = new NfeWebServices({
  uf: "DF",
  env: "homologacao",
  certificate: certificate,
});

const statusServ = await service.statusServico();
console.log(JSON.stringify(statusServ, null, 2));

const consCad = await service.consultaCadastro({
  CPF: process.env.TEST_CPF ?? "",
});
console.log(JSON.stringify(consCad, null, 2));

const consProt = await service.consultaProtocolo({
  chNFe: process.env.TEST_CHNFE ?? "",
});
console.log(JSON.stringify(consProt, null, 2));

const dfe = await service.distribuicaoDfe({
  CPF: process.env.TEST_CERTIFICATE_CPF ?? "",
  consChNFe: { chNFe: process.env.TEST_CHNFE ?? "" },
});
console.log(JSON.stringify(dfe, null, 2));
