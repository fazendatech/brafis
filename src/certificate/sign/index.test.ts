import { describe, expect, test } from "bun:test";
import { CertificateP12 } from "@/certificate";
import { signXml } from ".";

describe("signXml", async () => {
  const certificate = await CertificateP12.fromFilepath({
    filepath: "misc/sample-certificates/cert.pfx",
    password: "senha",
  });

  test("Signs XML", () => {
    const signId = "12345";
    const xmlObject = { root: { toSign: { "@_Id": signId } } };

    expect(signXml({ xmlObject, signId, certificate })).toMatchSnapshot();
  });

  test("Throws error when sign ID is not found", () => {
    const signId = "12345";
    const invalidSignId = "123456";
    const xmlObject = { root: { toSign: { "@_Id": signId } } };

    expect(() =>
      signXml({ xmlObject, signId: invalidSignId, certificate }),
    ).toThrowError(
      `the following xpath cannot be signed because it was not found: //*[@Id='${invalidSignId}']`,
    );
  });
});
