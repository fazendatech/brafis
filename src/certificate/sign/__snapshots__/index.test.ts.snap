// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`signXml Signs XML 1`] = `
{
  "root": {
    "Signature": {
      "@_xmlns": "http://www.w3.org/2000/09/xmldsig#",
      "KeyInfo": {
        "X509Data": {
          "X509Certificate": "MIIFeTCCA2GgAwIBAgIUc92p4Ro7Vbj/zCmuWLEyAOLJdcIwDQYJKoZIhvcNAQELBQAwUzEbMBkGA1UEAwwSZmF6ZW5kYXRlY2guY29tLmJyMREwDwYDVQQLDAhEZXYgVGVhbTEUMBIGA1UECgwLRmF6ZW5kYVRlY2gxCzAJBgNVBAYTAkJSMCAXDTI1MDEwODAyNTIyMVoYDzIwOTkxMjMxMDI1MjIxWjBTMRswGQYDVQQDDBJmYXplbmRhdGVjaC5jb20uYnIxETAPBgNVBAsMCERldiBUZWFtMRQwEgYDVQQKDAtGYXplbmRhVGVjaDELMAkGA1UEBhMCQlIwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIKAoICAQC+dPrxTGmuiEeFBXPTNy/zfZGycHgyNH8V5q7GAO2LlYIuxcHNrRUevZK8aFObxy6PDlJtjZNU5IgqDDUkokPv0ECHvQoG9hD/kcikWf8f0jfl7DqBRgyPC3NH2ihASyITIGhpb+3Rea7NKUQd/4xCMzsGdgH7Ywq/lNKsyE0ap9MhV+oxIXKD8mmYS+XSYKoNT2gakPLH+wMk4PqAiGtcypXKsVb9HkajOqAxt0F2OzXLg98yFB8iZdoR8Ia4kTD3CcI9ljV8uHxmLqHWIEfwGHHmHlKR0FFx6WGFwvlcG/ymUux2NBeE7Bb1Myte4dbeikXe7PsW2AWbgaVac1AJR+ZPvKbXWnV5rn3o/lHchW1Zz9npm1iWhNFb5SOSKykYYAj2YEg/2oJ0ozdG/TyKTIsSTVhDeepgEZS3O46qU4DxplwmfMpdRjskRlTwap4TzGfgBs+bL3dtOnJLvjms9gOAUlKcXrD85bvFMT4wlSG2CjWVa5O1db05qDqiN9SdUh5xzljR9e7aG8i2TmP3blblxrF0ineYZf6IRqtZFixWxlhkAzLzA9jbrdDL6D4Fzxy5ja5BTaG4y75QBhmPP84fL+zuOWdTS6g1Nice8fxnrhXMUO1cLu+0qsYG1eG6AWQhE/MJwp13/s0+GhWU1qJ3nKjSRqa9pr0MVDPX0QIDAQABo0MwQTALBgNVHQ8EBAMCBDAwEwYDVR0lBAwwCgYIKwYBBQUHAwEwHQYDVR0OBBYEFGC1E6b0Ya/4GhppHoddL+tlGjErMA0GCSqGSIb3DQEBCwUAA4ICAQA0GcWFIbuRO7+rr7M+np9jRrghvBVSyXWJoKryY5XvKeMrwg5VZROz84mz9nYng20hBD6SXi/h2l68Cg+R+Nvk5zTKbzLsh21/LxltjLdTWmtpwWtlLAhcJFpNUnbKJ/UNxOrn7rVCjbizzcjJmuZ8WExn8mgset1VZ1LDX1+06h2pV/SidbZy/JxEad02AabVq9hyx0WKAxLEt41abz7H/4MyPdggwxSSFQ2Y+UjtSeBnsx/iLbdxN0cQsaGW+FI+XZbapCNGXGgkrqOFxW6RLisMGWo013Jf7Y/YmMnTt3XBbntKn8WX46BIrQtrmtIfxG38QySj1RxHaFxL5MlioS7Il6LfjQMPcyOmoPx0+11bTKoZ1He6TEBMImewr6A94/TTOxAw5EbgwSNxYnRS0Z7wNTIe4/281TN0Y5AB8Bye8TJDlJB89Vq7ZUmhkVVQSFgFf6WPALLkesHTDP4qOrlPmb5l5fcy7cVM70pyW2PG/BrSmefWImHO9O8HAvfV1rr0QeU+y01uRLjBPhjux79+tw+tUg2p7JY0k+mwaizYzDM/cKq7gW6j4bBVO0HaTSLRpPm+MXmPJ5/UfeDHiXiZlsYzRIcKewJgpXA5uv/RErhZ78NlcUmqJT8PQL6PyjsIktResPV2ayACCYkWuW1jEIh9doqQpDpp0RcBnA==",
        },
      },
      "SignatureValue": "I2yGtMwvldf1l+Q5IDWgN1bvdrE9d8SZGbwGw1IbY6xE7IubugJfwMR0BYIoJbo+muNxZUOPHBnWGLzFPtivJ9DPhPPwBM7L/BgueXYdGyDtdWZodNWbOuVPTrzaYtmABvZ+YTB0wtpsheUzwYhKmj/JGcEVRyKFpOh7+JqaTp6M1CzlndVO//iCS9oeSsPki1G/mvrHx4nUHTJlOmajF0vtsH+fE5+IEWXEnbzT4Djc+pfAj6hVWtRB7yibLFTnjPaXlhombOFqbTg0T+vDdTcfdBPEaoHfKmPsF+3OXVoPQ7k9eSRwoqKhLiO21zgkGnIVykCcCVcCg5APZO9/9h7nisZT3pL+i1ujIUWkMVdG2yWq1Ngtrq0rhLCzKNF5WlAIR9R+JQ+QDBg3Q1lGZatQIXlLht1NJdhUKGBpmI2jvwdRXJdbEVrvOXy7hmlxvgB8mXO2qHQz3+TpJnlcPJjKFLIio8uQG/mFY2g6DHaDfWjcqA79OkWzpOC9/r7hkTtzv1JfZy3AZZiaMP7ZQYNQDteVVkk639TZgY5KeGkhQyo8Xd8xif7NcQEo7OiUQeCltF0xS/PAKC+/vpLw7hEyYpbG+FSNqydI7pGUMGSgIwX9XG8s8KMyFB47yhYjEV+4uY3iGmKxr9i0vlL3CFhE54csHqitFkpk8Umi25U=",
      "SignedInfo": {
        "CanonicalizationMethod": {
          "@_Algorithm": "http://www.w3.org/TR/2001/REC-xml-c14n-20010315",
        },
        "Reference": {
          "@_URI": "#12345",
          "DigestMethod": {
            "@_Algorithm": "http://www.w3.org/2000/09/xmldsig#sha1",
          },
          "DigestValue": "IUBQicjb6SEh+6Ubh2qtOWVXSYU=",
          "Transforms": {
            "Transform": [
              {
                "@_Algorithm": "http://www.w3.org/2000/09/xmldsig#enveloped-signature",
              },
              {
                "@_Algorithm": "http://www.w3.org/TR/2001/REC-xml-c14n-20010315",
              },
            ],
          },
        },
        "SignatureMethod": {
          "@_Algorithm": "http://www.w3.org/2000/09/xmldsig#rsa-sha1",
        },
      },
    },
    "toSign": {
      "@_Id": "12345",
    },
  },
}
`;
